feat: LEAP-1492: More flexible Custom buttons

hlomzik opened the pull request:
- extend the structure of custom buttons definition to allow different modes
- custom buttons now can replace just some buttons, leaving the rest untouched
- start to simplify work with all buttons to use standard button description + additional wrappers for main actions, so any button action will go through mandatory checks

Examples of new structure:
```js
// switch buttons
{ "_replace": ["update", "submit"] }

// add simple custom button
{ "_before": [{ "name": "analyze", "title": "Analyze task" }] }

// split Reject into 2 buttons, leaving normal Accept button;
// these 2 new buttons share part of reject logic
{ "reject": [
  { "name": "hide", "title": "Hide" },
  { "name": "postpone", "title": "Postpone" },
] }
```

### PR fulfills these requirements
- [ ] Tests for the changes have been added/updated (for bug fixes/features)
- [x] Docs have been added/updated (for bug fixes/features)
- [x] Best efforts were made to ensure docs/code are concise and coherent (checked for spelling/grammatical errors, commented out code, debug logs etc.)
- [x] Self-reviewed and ran all changes on a local instance (for bug fixes/features)

#### Change has impacts in these area(s)
- [x] Product design
- [ ] Backend (Database)
- [ ] Backend (API)
- [x] Frontend

### Describe the reason for change
We need to dynamically display buttons we need, so previous predefined approach didn't work.

#### Does this change affect performance?
no

#### Does this change affect security?
no

#### What alternative approaches were there?
_(briefly list any if applicable)_

### Does this PR introduce a breaking change?
- [ ] Yes, and covered entirely by feature flag(s)
- [ ] Yes, and covered partially by feature flag(s)
- [x] No
- [ ] Not sure (briefly explain the situation below)

### What level of testing was included in the change?
- [ ] e2e
- [ ] integration
- [ ] unit

hlomzik committed:
feat: LEAP-1492: More flexible Custom buttons

- extend the structure of custom buttons definition to allow different modes
- custom buttons now can replace just some buttons, leaving the rest untouched
- start to simplify work with all buttons to use standard button description + additional wrappers for main actions, so any button action will go through mandatory checks

Gondragos commented on web/libs/editor/src/components/BottomBar/Controls.tsx:
```diff
@@ -12,40 +12,44 @@ import { useCallback, useState } from "react";
 };
 
+/** If given one element, wrap it in an array */
+function toArray<T>(arg: undefined | T | (T | undefined)[]): T[] {
+  return (Array.isArray(arg) ? arg : [arg]).filter(v => v !== undefined);
+}
```
It could be moved to utils

hlomzik replied:
moved and added tests, thank you!

Gondragos commented on web/libs/editor/src/components/BottomBar/Controls.tsx:
```diff
@@ -12,40 +12,44 @@ import { useCallback, useState } from "react";
-    await onClick?.(button.name);
-    setWaiting(false);
-  }, []);
+  // @todo do we need waiting? all buttons should utilize isSubmitting
```
It's for indication, but it may look not good.

Gondragos commented on web/libs/editor/src/components/BottomBar/Controls.tsx:
```diff
@@ -98,29 +104,65 @@ export const Controls = controlsInjector<{ annotation: MSTAnnotation }>(
+
+    // either we render _before buttons and then the rest, or we render only _replace buttons
+    if (firstToRender) {
+      const allButtons = Array.isArray(firstToRender) ? firstToRender : [firstToRender];
```
```suggestion
      const allButtons = toArray(firstToRender);
```

Gondragos commented on web/libs/editor/src/components/BottomBar/Controls.tsx:
```diff
@@ -98,29 +104,65 @@ export const Controls = controlsInjector<{ annotation: MSTAnnotation }>(
+      const allButtons = Array.isArray(firstToRender) ? firstToRender : [firstToRender];
+      for (const customButton of allButtons) {
         // @todo make a list of all internal buttons and use them here to mix custom buttons with internal ones
-        if (customButton.name === "accept") {
-          buttons.push(<AcceptButton disabled={disabled} history={history} store={store} />);
```
Do we need a backward capability or not?

Gondragos commented on web/libs/editor/src/stores/types.d.ts:
```diff
@@ -29,7 +29,8 @@ type MSTCommentStore = {
 type MSTStore = {
-  customButtons: Instance<typeof CustomButton>[];
+  // @todo we can't import CustomButton store here and use it type :(
+  customButtons: any;
```
I've just figured out that it works for me 🤯
```suggestion
  customButtons: CustomControlProps.button[];
```

hlomzik replied:
that feels a bit weird because it will grab a type from js file which is also based on this definition. I'd leave it like this for now until we find a better way to define MST types

hlomzik committed:
Pass all properties to LSF coming from external sources

DM should pass all the options down to LSF, so it will be controllable.

yyassi-heartex commented on web/libs/editor/src/components/BottomBar/Controls.tsx:
```diff
@@ -12,40 +12,44 @@ import { useCallback, useState } from "react";
-    setWaiting(false);
-  }, []);
+  // @todo do we need waiting? all buttons should utilize isSubmitting
+  // const [waiting, setWaiting] = useState(false);
```
```suggestion
```

yyassi-heartex commented on web/libs/editor/src/components/BottomBar/Controls.tsx:
```diff
@@ -12,40 +12,44 @@ import { useCallback, useState } from "react";
-        onClick={clickHandler}
-        waiting={waiting}
+        onClick={onClick}
+        // waiting={waiting}
```
changing my initial suggestion - since we have isSubmitting, we should listen to that as you mentioned in your comment above. just gotta make sure that store is accessible here
```suggestion
        waiting={store.isSubmitting}
```

hlomzik replied:
Waiting has different style and is very helpful, because it indicates that something's happening and setting it only on one button will indicate which button was pressed. All buttons are disabled to prevent wrong moves, but the current one also has animation.
But it doesn't work well, because `onClick` handler send as event and apparently finishes too fast, so indication is broken. So I removed it now completely including comments. Will add later some day.

yyassi-heartex commented on web/libs/editor/src/components/BottomBar/Controls.tsx:
```diff
@@ -98,29 +104,65 @@ export const Controls = controlsInjector<{ annotation: MSTAnnotation }>(
-        if (customButton.name === "accept") {
-          buttons.push(<AcceptButton disabled={disabled} history={history} store={store} />);
+        // string buttons is a way to render internal buttons
+        if (typeof customButton === "string") {
+          if (customButton === "accept") {
```
what happens if we have a string customButton that is not "accept" - not sure if its possible but reading this indicates that we would not add a button

hlomzik replied:
next line says that this is just an example. just some buttons can be reused, for now it's only `accept` and `reject`

yyassi-heartex commented on web/libs/editor/src/components/BottomBar/Controls.tsx:
```diff
@@ -12,40 +12,44 @@ import { useCallback, useState } from "react";
       <Button
         aria-label={button.ariaLabel}
-        disabled={button.disabled || disabled || waiting}
+        disabled={button.disabled || disabled}// || waiting}
```
```suggestion
        disabled={button.disabled || disabled}
```

yyassi-heartex reviewed (approved):
looks mostly good - there is a couple of notes here and there and some replacement suggestions but non-blocking

hlomzik committed:
Remove commented out `waiting`, it's not usable right now

Button's handler is executed too fast, so `waiting` doesn't wait for the end of actual request.

hlomzik committed:
Fix crash with new comment format

`trim is not a function` because new comments are objects, not strings

hlomzik committed:
Fix init with custom buttons in old format

It was just an array and it's a map now. Auto convert array to _replace.

hlomzik commented:
/git merge
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/11163045587)
> Successfully merged:  35 files changed, 98 insertions(+), 56 deletions(-)

hlomzik committed:
Merge branch 'develop' into 'fb-leap-1492/more-flexible-buttons'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/11163045587

hlomzik committed:
reuse `toArray()`

Co-authored-by: Sergey <sergey.koshevarov@humansignal.com>

bmartel commented on web/libs/editor/src/components/BottomBar/Controls.tsx:
```diff
@@ -98,29 +104,65 @@ export const Controls = controlsInjector<{ annotation: MSTAnnotation }>(
-    // custom buttons replace all the internal buttons, but they can be reused if `name` is one of the internal buttons
-    if (store.customButtons?.length) {
-      for (const customButton of store.customButtons ?? []) {
+    const buttonsBefore = customButtons.get("_before");
+    const buttonsReplacement = customButtons.get("_replace");
```
```
    const buttonsBefore = customButtons.get("_before");
    const buttonsReplacement = customButtons.get("_replace");
```
Why the difference in names, these ones have `_` but all others are just normal keys in the map. Is this supposed to be a private API?

hlomzik replied:
This way I'm trying to separate internal buttons (like `accept` and `reject`) and special places (`_before`, `_after`, `_replace`). I'll share a document soon.

bmartel commented on web/libs/editor/src/components/BottomBar/Controls.tsx:
```diff
@@ -12,40 +12,41 @@ import { useCallback, useState } from "react";
-  button: Instance<typeof CustomButton>;
+type CustomButtonType = Instance<typeof CustomButton>;
+// @todo should be Instance<typeof AppStore>["customButtons"] but it doesn't fit to itself
+type CustomButtonsField = Map<string, CustomButtonType | string | Array<CustomButtonType | string>>;
```
If the map keys are well-known, we should just list them out so there are some types forming a governance to how this is expected to be used externally. That way without reading the rest of the file a reader would know the important keys that have more meaning in the implementation.
```
type CustomButtonsFieldKeys = "accept" | "reject" | string;
type CustomButtonsField = Map<CustomButtonsFieldKeys, CustomButtonType | string | Array<CustomButtonType | string>>;
```

hlomzik replied:
thank you, made it more clear + comments

hlomzik committed:
Add stub controlButtons to fix Controls.test.tsx

hlomzik committed:
Fix linting

hlomzik committed:
Move `toArray()` to utilities

We already have `wrapArray()` but it has broken types and a little different semantic

hlomzik committed:
Add comments and types for string keys in custom buttons
